permits/geocode.py

```diff
@@ -0,0 +1,188 @@
+        return {
+            'geocoded_address': address,
+            'geocoder': 'coa_feature_server',
+            'lat': lat,
+            'lng': lng,
+            'city': city_name,
+            'postal_code': zipcode,
+            'state': 'Texas',
+        }
+
+
+def geocode_from_mapzen(permit_location):
+    address = permit_location.upper().strip()
+    for scrubber in SCRUBBERS:
+        address = address.replace(scrubber, ' ')
+    address = '{}, Austin, TX'.format(address)
+
+    geocode = geocoder.mapzen(address, key=secrets.MAPZEN_API_KEY)
+    geocoded_feature = {
+        'geocoded_address': geocode.address,
+        'geocoder': 'mapzen',
+        'lat': geocode.lat,
+        'lng': geocode.lng,
+        'accuracy': geocode.accuracy,
+        'city': geocode.city,
+        'county': getattr(geocode, 'county', ''),
+        'state': geocode.state,
+        'postal_code': geocode.postal,
+    }
+    return geocoded_feature
+
+
+def city_name_for_point(lng, lat):
+    """returns city name for a given point, queried from CoA jurisdiction
+    boundary data"""
+    url = 'http://services.arcgis.com/0L95CJ0VTaxqcmED/ArcGIS/rest/services/BOUNDARIES_jurisdictions/FeatureServer/0/query'
+    return _property_where_intersects(url, 'CITY_NAME', lng, lat)
+
+
+def zipcode_for_point(lng, lat):
+    """returns zipcode for a given point, queried from CoA zipcode boundary data"""
+    url = 'http://services.arcgis.com/0L95CJ0VTaxqcmED/ArcGIS/rest/services/LOCATION_zipcodes/FeatureServer/0/query'
+    return _property_where_intersects(url, 'ZIPCODE', lng, lat)
+
+
+def _get_single_feature_only(json_response):
+    """returns feature from ArcGIS feature server geojson response only when
+    there is a single feature in the response. Returns None if there are no
+    features or more than one feature.
+    """
+    features = json_response.get('features')
+    if features is not None and len(features) != 1:
+        return
+    else:
+        return features[0]
+
+
+def _property_where_intersects(query_url, property_name, lng, lat):
+    """queries for feature that intersects a given point, an returns the value
+    of property_name at that point"""
+    req = requests.get(query_url, {
+        'geometry': '{lng},{lat}'.format(lng=lng, lat=lat),
+        'geometryType': 'esriGeometryPoint',
+        'inSR': '4326',
+        'spatialRel': 'esriSpatialRelIntersects',
+        'outFields': '*',
+        'returnGeometry': 'false',
+        'outSR': '4326',
+        'f': 'pgeojson'
+    })
+    json_response = req.json()
+    feature = _get_single_feature_only(json_response)
+    if feature:
+        return feature['properties'].get(property_name)
+    else:
+        return ''
+
+
+def _query_coa_geocoder(street_name):
+    url = "http://www.austintexas.gov/GIS/REST/Geocode/COA_Composite_Locator/GeocodeServer/findAddressCandidates"
+    req = requests.get(url, {
+        'street': street_name,
+        'f': 'pjson',
```

Comment:
you can add `'outSR': '4326'` to get lat/long back from the geocoding service itself and skip the clientside reprojection.

this would be *a lot* easier to see in the REST service landing page if it didn't try to redirect you to an invalid URL, but you can see an example [here](http://www.austintexas.gov/GIS/REST/Geocode/COA_Composite_Locator/GeocodeServer/findAddressCandidates?f=pjson&singleLine=main&outSR=4326).